File: contest/LeetCode_Weekly_Contest_201/Minimum_Cost_to_Cut_a_Stick.java
class Solution {
    int dfs(int[] cuts, int[][] dp, int n, int m, int st, int ed) {
        if (ed < st) return 0;
        if (dp[st][ed] != -1) {
            return dp[st][ed];
        }

        int L = 0, R = n;
        if (st != 0) L = cuts[st-1];
        if (ed != m-1) R = cuts[ed+1];

        int val = R - L;
        int ret = Integer.MAX_VALUE;
        dp[st][ed] = ret;
        for (int i = st; i <= ed; ++i) if (L < cuts[i] && cuts[i] < R) {
            ret = Math.min(ret,
                    dfs(cuts, dp, n, m, st, i-1) +
                    dfs(cuts, dp, n, m, i+1, ed) +
                    val);
        }

        if (ret == Integer.MAX_VALUE) ret = 0;
        dp[st][ed] = ret;
        return ret;
    }

    public int minCost(int n, int[] cuts) {
        Arrays.sort(cuts);
        int m = cuts.length;
        int[][] dp = new int[m+1][m+1];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < m; ++j) {
                dp[i][j] = -1;
            }
        }

        return dfs(cuts, dp, n, m,0, m-1);
    }
}

File: contest/LeetCode_Weekly_Contest_201/Minimum_Cost_to_Cut_a_Stick.js
/**
 * @param {number} n
 * @param {number[]} cuts
 * @return {number}
 */
var minCost = function(n, cuts) {
    cuts.sort(function(a, b) { return a - b; });
    var m = cuts.length;
    var dp = [];
    for (var i = 0; i <= m; ++i) {
        dp[i] = [];
        for (var j = 0; j <= m; ++j)
            dp[i][j] = -1;
    }
    
    var dfs = function(st, ed) {
        if (ed < st) return 0;
        if (dp[st][ed] != -1) return dp[st][ed];
        
        var L = 0, R = n;
        if (st != 0) L = cuts[st-1];
        if (ed != m-1) R = cuts[ed+1];
        
        var ret = Number.MAX_SAFE_INTEGER;
        dp[st][ed] = ret;
        var val = R - L;
        for (var i = st; i <= ed; i++) if (L < cuts[i] && cuts[i] < R) {
            ret = Math.min(ret, dfs(st, i-1) + dfs(i+1, ed) + val);
        }
        if (ret === Number.MAX_SAFE_INTEGER) 
            ret = 0;
        dp[st][ed] = ret;
        return ret;
    };
    
    return dfs(0, m-1);
};
